### PR TITLE
Updated to include archive functionality

### PR DESCRIPTION
**Pull request by ccloes**
Work in progress.  Adding tests now.

### CODE REVIEW

File: nodestream/extractors/stores/aws/s3_extractor.py [hunk 1]
```diff
@@ -55,10 +70,17 @@ def find_keys_in_bucket(self) -> list[str]:
...
+                    yield obj["Key"]
 
     async def extract_records(self) -> AsyncGenerator[Any, Any]:
         for key in self.find_keys_in_bucket():
             for record in self.get_object_as_file(key).read_file():
-                yield record
+                try:
+                    yield record
+                finally:
+                    self.archive_s3_object(key)
```
Comment (zprobst) [inline nodestream/extractors/stores/aws/s3_extractor.py:86]:
```python
for key in self.find_keys_in_bucket():
    try:
        for record in self.get_object_as_file(key).read_file():
            yield record
    finally:
        self.archive_s3_object(key)
```

I think we need to move the `try-finally` block out of the loop so it doesn't try to archive on every record in the file.

File: nodestream/extractors/stores/aws/s3_extractor.py [hunk 2]
```diff
@@ -14,30 +14,45 @@ def from_file_data(
...
         self.bucket = bucket
+        self.archive_dir = archive_dir
         self.s3_client = s3_client
 
     def get_object_as_io(self, key: str) -> StringIO:
         return self.s3_client.get_object(Bucket=self.bucket, Key=key)["Body"]
 
+    def archive_s3_object(self, key: str):
+        if self.archive_dir:
+            self.logger.info("Archiving S3 Object", extra=dict(key=key))
```
Comment (zprobst) [inline nodestream/extractors/stores/aws/s3_extractor.py:47]: I don't think currently `S3Extractor` has a logger in here currently. I think its appropriate to add this one.

File: nodestream/extractors/stores/aws/s3_extractor.py [hunk 3]
```diff
@@ -14,30 +14,45 @@ def from_file_data(
         bucket: str,
         prefix: Optional[str] = None,
         object_format: Optional[str] = None,
+        archive_dir: str = None,
```
Comment (zprobst) [inline nodestream/extractors/stores/aws/s3_extractor.py:17]: technically this is optional

```suggestion
        archive_dir: Optional[str] = None,
```

File: nodestream/extractors/stores/aws/s3_extractor.py [hunk 4]
```diff
@@ -14,30 +14,45 @@ def from_file_data(
...
             object_format=object_format,
             prefix=prefix,
+            archive_dir=archive_dir,
             s3_client=AwsClientFactory(**aws_client_args).make_client("s3"),
         )
 
     def __init__(
         self,
         bucket: str,
+        archive_dir: str,
```
Comment (zprobst) [inline nodestream/extractors/stores/aws/s3_extractor.py:31]:
```suggestion
        archive_dir:  Optional[str],
```
Same thing as above.

File: nodestream/extractors/stores/aws/s3_extractor.py (hunk 1 above)
Comment (ccloes) [reply nodestream/extractors/stores/aws/s3_extractor.py]: DOH... i think i fixed it now

File: nodestream/extractors/stores/aws/s3_extractor.py (hunk 2 above)
Comment (ccloes) [reply nodestream/extractors/stores/aws/s3_extractor.py:49]: Done

File: nodestream/extractors/stores/aws/s3_extractor.py (hunk 3 above)
Comment (ccloes) [reply nodestream/extractors/stores/aws/s3_extractor.py]: Done.  Please double check me

File: nodestream/extractors/stores/aws/s3_extractor.py (hunk 4 above)
Comment (ccloes) [reply nodestream/extractors/stores/aws/s3_extractor.py]: Same.  Done, please double check me

**Pull request merged**